internal/zoneinfo/zoneinfo_clock.go

```diff
@@ -0,0 +1,57 @@
+package zoneinfo
+
+import "time"
+
+// ToLocalClock transposes the utc timestamp from a time in the UTC
+// location to an equivalent clock time in the current location.
+func (l *Location) ToLocalClock(utc int64) (local int64) {
+	// This function does its math in seconds rather than nanoseconds.
+	sec := utc / int64(time.Second)
+
+	// Look for zone offset for sec, so we can adjust from UTC.
+	// The lookup function expects the local time, so we pass sec in the
+	// hope that it will not be too close to a zone transition.
+	// If the adjusted time is not in the zone transition,
+	// then we check the previous or next zone transition to see if
+	// it is the appropriate one.
+	// If there is no appropriate zone transition, the clock time does not
+	// exist in that location and we truncate the time to the start or end
+	// of the zone transition.
+	_, offset, start, end, _ := l.lookup(sec)
+	if offset != 0 {
+		// The time may not be valid in our present zone offset.
+		// If it is not, check to see if the previous or next zone offset
+		// would be applicable.
+		switch local := sec - int64(offset); {
+		case local < start:
+			_, offset, start, end, _ = l.lookup(start - 1)
+			if local := sec - int64(offset); local < start || local >= end {
+				// The time does not appear to exist. Truncate the interval.
+				return end * int64(time.Second)
+			}
+		case local >= end:
+			_, offset, start, end, _ = l.lookup(end)
+			if local := sec - int64(offset); local < start || local >= end {
+				// The time does not appear to exist. Truncate the interval.
+				return start * int64(time.Second)
+			}
+		}
+	}
+
+	// Inspect the previous time zone transition.
```

Comment:
I know we've talked about this in meetings before, but I think I'm still unclear on what the issue is with the go time library that makes this code necessary. I don't doubt the logic here, I think I would just like to understand the problem it's trying to solve in greater detail.

Reply:
I'll expand this comment. The main reason is that we're essentially guessing which zone transition, which is a start and end timestamp along with an offset, that applies to the time we want. Since there are possibly multiple times that correspond to a single utc clock time, we need a way to consistently determine which of them we want. We've decided to say "the first clock time that occurs is the correct one" which means that the time we are looking for is in the earlier zone transition. Since the above code guesses the zone transition by looking it up with the utc timestamp (which isn't correct), we need to check to see if there is another zone transition that could also be the correct one. Because we only care about this condition to find the earliest timestamp, we only have to check the previous one. We don't really care if we accidentally chose the first one and there's also an available zone transition in the future because we will never choose it anyway.